src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs

```diff
@@ -548,7 +548,7 @@ private Task WaitForExitAsync(CancellationToken cancellationToken = default)
                         _waitInProgress = null;
                     }
                 }
-            }, cancellationToken);
+            }, CancellationToken.None); // We should avoid this Task.Run being cancelled and the delegate being skipped
```

Comment:
An async method (or an sync wrapper on async) could take more time than expected to finish when thread pool queue is backed up. I suppose it's acceptable. 

Even if this Task.Run could be canceled after `remainingTimeout` I think there's still a problem: it takes `lock(_gate)` for too much time.

Reply:
> An async method (or a sync wrapper on async) could take more time than expected to finish

To finish, yes.  To be canceled, the goal is that's much more prompt.  The whole point of cancellation is to say "I no longer need this anymore, I'd like to go do something else", and we want to avoid stalling such completions for arbitrarily long periods of time.

Reply:
> we should be handling the Wait more gracefully and translating a completion as canceled into a timeout

Can this be achieved without using exceptions?

```cs
// Polling loop
while (true)
{
    try
    {
        waitTask.Wait();
    }
    catch (AggregateException ae) when (ae.InnerException is TimeoutException)
    {
        continue;
    }
}
```

Reply:
It seems that `Task.WaitAny` could be used but we need a proper way to handle `_waitInProgress`.